### PR TITLE
[SPARK-27011][SQL] reset command fails with cache

### PR DESCRIPTION
**Pull request by ajithme**
## What changes were proposed in this pull request?

When cache is enabled ( i.e once cache table command is executed), any following sql will trigger 
 CacheManager#lookupCachedData which will create a copy of the tree node, which inturn calls TreeNode#makeCopy. Here the problem is it will try to create a copy instance. But as ResetCommand is a case object this will fail

## How was this patch tested?

Added UT to reproduce the issue

### CODE REVIEW

**Comment by ajithme** [comment]
any suggestions @gatorsmile @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 1]
```diff
@@ -377,8 +377,7 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
    * @param newArgs the new product arguments.
    */
   def makeCopy(newArgs: Array[AnyRef]): BaseType = attachTree(this, "makeCopy") {
-    // Skip no-arg constructors that are just there for kryo.
-    val ctors = getClass.getConstructors.filter(_.getParameterTypes.size != 0)
+    val ctors = getClass.getConstructors
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: Why this change?

Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: As now case class (ResetCommand) has a zero parameter constructor, we cannot filter it out

**Comment by srowen** [comment]
That looks plausible and does seem like a real problem. @gatorsmile does that look right to you?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala (hunk 1 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: Simply removing this does not sound right to me. We need to double check the impact. cc @cloud-fan

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: See https://github.com/apache/spark/pull/5157/files

**Comment by gatorsmile** [comment]
This is interesting. We also can fix the issue by avoiding cache lookup for RESET, which will not use CACHE anyway.

**Comment by ajithme** [comment]
Thanks for the inputs @srowen @gatorsmile , i have updated the PR accordingly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala (hunk 1 above)
Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: @gatorsmile i have changed the fix to override makeCopy in ResetCommand instead now

**Comment by ajithme** [comment]
Thanks for the inputs @srowen @gatorsmile , i have updated the PR accordingly.

**Comment by ajithme** [comment]
@srowen @gatorsmile @vanzin , i have updated the PR accordingly. Please review

**Comment by cloud-fan** [comment]
+1 on @gatorsmile 's idea. We shouldn't lookup cache for commands like `ResetCommand`.

**Comment by ajithme** [comment]
@gatorsmile @cloud-fan @srowen So now i have overridden the makecopy() in ResetCommand ( just like ClearCacheCommand), would that suffice for this bug or should i change so that CacheManager avoids a lookup call in case of (ResetCommand, ClearCacheCommand etc).??
Will latter approach be beyond the scope of this jira.? please suggest

**Comment by cloud-fan** [comment]
Yea the later approach is beyond the scope of this JIRA, but it's also a superior solution to this problem. We can create a trait `IgnoreCachedData` and let `ResetCommand` extends it.

**Comment by ajithme** [comment]
@cloud-fan @srowen @gatorsmile i have updated the PR to use the new approach as solution of using ``IgnoreCachedData`` trait. Please review

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 2]
```diff
@@ -198,3 +198,6 @@ case class StreamingExplainCommand(
     ("Error occurred during query planning: \n" + cause.getMessage).split("\n").map(Row(_))
   }
 }
+
+/** Mark that the command can be skipped from cache lookup */
+trait IgnoreCachedData {}
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: can we move it to catalyst module and make it more general? I'm not sure if only commands need it
```
/**
 * A [[LogicalPlan]] operator that does not use the cached results stored in CacheManager.
 */
trait IgnoreCachedData extends LogicalPlan

```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 3]
```diff
@@ -245,7 +246,7 @@ class CacheManager extends Logging {
       // node, so that the returned caching plan won't replace the hint node and drop the hint info
       // from the original plan.
       case hint: ResolvedHint => hint
-
+      case command: IgnoreCachedData => command
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: nit: move it to the first case so that it's more obvious we ignore it
```
plan transformDown {
  case p: IgnoreCachedData => p
  ...
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 2 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Moved to catalyst

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala (hunk 3 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: Agreed, reordered match-case

**Comment by ajithme** [comment]
@cloud-fan i have updated as per review comments. Please review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Command.scala [hunk 4]
```diff
@@ -28,3 +28,6 @@ trait Command extends LogicalPlan {
   override def output: Seq[Attribute] = Seq.empty
   override def children: Seq[LogicalPlan] = Seq.empty
 }
+
+/** Mark that the command can be skipped from cache lookup */
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Command.scala]: yes, you moved it to catalyst. Can we make it more general then? I think it's not only commands that can skip cache lookup.

Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Command.scala]: I have moved it to a diff scala file, would that suffice.?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/cache.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+/** Mark that the CacheManager can skip this instance */
+trait IgnoreCachedData {}
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/cache.scala]: we only have a single class in this file, we can rename the file to `IgnoreCachedData.scala`

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/cache.scala]: we can also improve the comments
```
/**
 * A [[LogicalPlan]] operator that does not use the cached results stored in CacheManager.
 */
trait IgnoreCachedData extends LogicalPlan
```

Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/cache.scala]: Updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IgnoreCachedData.scala [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+/**
+ * A [[LogicalPlan]] operator that does not use the cached results stored in CacheManager
+ */
+trait IgnoreCachedData {}
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IgnoreCachedData.scala]: super nit: `extends LogicalPlan`

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IgnoreCachedData.scala (hunk 6 above)
Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IgnoreCachedData.scala]: fixed

**Comment by ajithme** [comment]
can test build be triggered for this.??

**Review by maropu — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IgnoreCachedData.scala [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+/**
+ * A [[LogicalPlan]] operator that does not use the cached results stored in CacheManager
+ */
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IgnoreCachedData.scala]: Could you put comments about which operator should use this trait? This is like `If a logical plan is a case object, you should use this trait...`.

Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IgnoreCachedData.scala]: @maropu even a non case logical plan if needed can be forced not to use cache. Right.?? So is the comment really needed.?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IgnoreCachedData.scala]: Never mind. Its ok as it is.

**Review by maropu — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 8]
```diff
@@ -3003,6 +3003,16 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test("reset ") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: nit: `reset command should not fail with cache`

Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Done.

**Comment by ajithme** [comment]
@maropu I have updated the test case name as per your suggestion. Please review

**Comment by maropu** [comment]
LGTM, pending Jenkins

**Comment by maropu** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!
